runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/Document.kt

```diff
@@ -14,6 +14,17 @@ sealed class Document {
      * Wraps a [kotlin.Number] of arbitrary precision.
      */
     data class Number(val value: kotlin.Number) : Document() {
+        init {
+            if (value == Double.POSITIVE_INFINITY ||
+                value == Double.NEGATIVE_INFINITY ||
+                value == Double.NaN
+            ) {
+                throw IllegalArgumentException(
+                    "a document number cannot be $value, as its value cannot be preserved across serde"
+                )
+            }
```

Comment:
**Comment**: Missing checks for the `Float` equivalents of these. The `isFinite()` functions could probably help with brevity. Might be able to clarify this as:

```kotlin
if (value is Double && !value.isFinite() || value is Float && !value.isFinite()) {
    …
}
```